--- linux-3.16.3-clean/kernel/time/timekeeping.c	2014-09-17 11:22:16.000000000 -0600
+++ ./linux-3.16.3-vtmininet/kernel/time/timekeeping.c	2017-09-01 12:53:48.450976718 -0600
@@ -19,6 +19,7 @@
 #include <linux/clocksource.h>
 #include <linux/jiffies.h>
 #include <linux/time.h>
+#include <uapi/linux/time.h>
 #include <linux/tick.h>
 #include <linux/stop_machine.h>
 #include <linux/pvclock_gtod.h>
@@ -290,6 +292,26 @@ static void timekeeping_forward_now(stru
 	timespec_add_ns(&tk->raw_time, nsec);
 }
 
+/* Delta virtual time = delta physical time / TDF */
+static void do_dilatetimeofday(struct timespec *ts) {
+	if (current->dilation > 0 && current->virtual_start_nsec > 0) {
+		u64 now = timespec_to_ns(ts);
+		u64 dilated_now = current->virtual_start_nsec + 
+			(now - current->virtual_start_nsec) / current->dilation;
+
+		if (dilated_now > now) {
+			printk("[panic] [process %d] VT (%lld) faster than RT (%lld) with dilation = %d",
+				current->pid, dilated_now, now, current->dilation);
+			return;
+		}
+
+		struct timespec dilated_ts = ns_to_timespec(dilated_now);
+
+		ts->tv_sec = dilated_ts.tv_sec;
+		ts->tv_nsec = dilated_ts.tv_nsec;
+	}
+}
+
 /**
  * __getnstimeofday - Returns the time of day in a timespec.
  * @ts:		pointer to the timespec to be set
@@ -320,6 +349,10 @@ int __getnstimeofday(struct timespec *ts
 	 */
 	if (unlikely(timekeeping_suspended))
 		return -EAGAIN;
+
+	/* Dilate time */
+	do_dilatetimeofday(ts);
+
 	return 0;
 }
 EXPORT_SYMBOL(__getnstimeofday);
