--- linux-3.16.3-clean/net/sched/sch_htb.c	2014-09-17 11:22:16.000000000 -0600
+++ ./linux-3.16.3-vtmininet/net/sched/sch_htb.c	2017-08-29 10:02:36.912974066 -0600
@@ -37,6 +37,8 @@
 #include <linux/rbtree.h>
 #include <linux/workqueue.h>
 #include <linux/slab.h>
+#include <linux/sched.h> // for accessing current
+
 #include <net/netlink.h>
 #include <net/sch_generic.h>
 #include <net/pkt_sched.h>
@@ -97,6 +100,7 @@ struct htb_prio {
 struct htb_class {
 	struct Qdisc_class_common common;
 	struct psched_ratecfg	rate;
+	int dilation;
 	struct psched_ratecfg	ceil;
 	s64			buffer, cbuffer;/* token bucket depth/rate */
 	s64			mbuffer;	/* max wait time */
@@ -857,6 +861,15 @@ next:
 
 	} while (cl != start);
 
+	/* polling host process about dilation info; update rate when necessary */
+	if (current) {
+		/* get tdf, possibly an updated one */
+	    if (current->dilation != cl->dilation) {
+		   cl->dilation = current->dilation; // update dilation
+		   psched_ratecfg_dilate(&cl->rate, cl->dilation); // update rate
+	    }
+   }
+
 	if (likely(skb != NULL)) {
 		bstats_update(&cl->bstats, skb);
 		cl->un.leaf.deficit[level] -= qdisc_pkt_len(skb);
